subprojects/gradle/runner/service/src/main/kotlin/com/avito/runner/service/listener/TestMetricsSender.kt

```diff
@@ -0,0 +1,81 @@
+package com.avito.runner.service.listener
+
+import com.avito.android.stats.GaugeMetric
+import com.avito.android.stats.StatsDSender
+import com.avito.logger.LoggerFactory
+import com.avito.logger.create
+import com.avito.runner.service.model.TestCase
+import com.avito.runner.service.model.TestCaseRun
+import com.avito.runner.service.worker.device.Device
+import com.avito.time.TimeProvider
+
+internal class TestMetricsSender(
+    private val statsDSender: StatsDSender,
+    private val timeProvider: TimeProvider,
+    loggerFactory: LoggerFactory
+) : TestListener {
+
+    private val prefix = "runner.test.duration"
+
+    private val logger = loggerFactory.create<TestMetricsSender>()
+
+    private val intendedMarks = mutableMapOf<Key, Long>()
+
+    private val startedMarks = mutableMapOf<Key, Long>()
+
+    override fun intended(test: TestCase, targetPackage: String, executionNumber: Int) {
+        intendedMarks[Key(test, executionNumber)] = timeProvider.nowInMillis()
+    }
+
+    override fun started(device: Device, targetPackage: String, test: TestCase, executionNumber: Int) {
+        val key = Key(test, executionNumber)
+        startedMarks[key] = timeProvider.nowInMillis()
+
+        val intendedMark = intendedMarks[key]
+
+        if (intendedMark != null) {
+            statsDSender.send(
+                prefix = prefix,
+                metric = GaugeMetric(
+                    path = keyMetric("queue", key),
+                    value = timeProvider.nowInMillis() - intendedMark
+                )
+            )
+        } else {
+            logger.warn("Can't send test metric: intended mark not found")
+        }
+    }
+
+    override fun finished(
+        device: Device,
+        test: TestCase,
+        targetPackage: String,
+        result: TestCaseRun.Result,
+        durationMilliseconds: Long,
+        executionNumber: Int
+    ) {
+        val key = Key(test, executionNumber)
+
+        val startedMark = startedMarks[key]
+
+        if (startedMark != null) {
+            statsDSender.send(
+                prefix = prefix,
+                metric = GaugeMetric(
+                    path = keyMetric("execution", key),
+                    value = timeProvider.nowInMillis() - startedMark
+                )
+            )
+        } else {
+            logger.warn("Can't send test metric: started mark not found")
+        }
+    }
+
+    private fun keyMetric(name: String, key: Key): String =
+        "${key.test.deviceName}.${key.executionNumber}.${key.test.testName}.$name"
```

Comment:
Let's make the format "testName.deviceName.executionNumber.metricName"

Reply:
As I said it will produce a lot of unique metrics and it will load graphite a lot.

As I understand we don't need specific test data we need only average time for all tests

Let's simplify metric sending and remove testName, device and execution number and Statsd will produce a unique point for each 30s